File: scripts/scaffolds/mlr/inference.py
#!/usr/bin/env python3
"""
Multiple Linear Regression - Remote Inference Script
Performs prediction using trained MLR model parameters.
**CRITICAL**: Model was trained on PERCENTAGE RETURNS with 2 lags.
Used for server-side scheduled predictions (Lambda).
Input: parameters.json, price data
Output: Price predictions
"""

import json
import sys
import numpy as np
from typing import Dict, Any, List


def predict(prices: List[float], parameters: Dict[str, Any]) -> List[Dict[str, float]]:
    """
    Make predictions using the trained MLR model.
    **Model trained on % returns**: Input = [return[t-1], return[t-2]], Output = return[t]
    
    Args:
        prices: List of historical prices
        parameters: Trained model parameters (coefficients, intercept)
    
    Returns:
        List of predictions with price and predicted next price
    """
    coefficients = np.array(parameters['coefficients'])
    intercept = parameters['intercept']
    
    # Convert prices to percentage returns
    prices_array = np.array(prices)
    returns = (prices_array[1:] / prices_array[:-1] - 1) * 100
    
    predictions = []
    
    # Predict next return using 2 lagged returns
    # Need at least 3 prices (2 returns) to make first prediction
    for i in range(2, len(returns)):
        current_price = prices[i + 1]
        
        # Features: [return[t-1], return[t-2]]
        X = np.array([returns[i - 1], returns[i - 2]])
        
        # Predict next return
        predicted_next_return = intercept + X @ coefficients
        
        # Convert predicted return back to price
        predicted_next_price = current_price * (1 + predicted_next_return / 100)
        
        predictions.append({
            'close': float(current_price),
            'close_pred': float(predicted_next_price),
            'return_t-1': float(returns[i - 1]),
            'return_t-2': float(returns[i - 2]),
            'return_pred': float(predicted_next_return)
        })
    
    return predictions


if __name__ == '__main__':
    """
    Command-line interface for inference.
    Expected input (stdin): JSON object with 'parameters' and 'data' keys
    Output (stdout): JSON array of predictions
    """
    try:
        # Read input from stdin
        input_data = json.loads(sys.stdin.read())
        
        # Validate input
        if 'parameters' not in input_data:
            raise ValueError("Input must contain 'parameters' key")
        if 'data' not in input_data:
            raise ValueError("Input must contain 'data' key")
        
        parameters = input_data['parameters']
        data = input_data['data']
        
        # Extract prices from data (use 'close' field, not the synthetic return fields from training)
        try:
            prices = [float(row['close']) for row in data]
        except (KeyError, ValueError) as e:
            raise ValueError(f"Error extracting prices from 'close' field: {e}")
        
        # Make predictions
        predictions = predict(prices, parameters)
        
        # Output predictions as JSON
        print(json.dumps(predictions, indent=2))
        sys.exit(0)
        
    except Exception as e:
        error_output = {
            'error': str(e),
            'type': type(e).__name__
        }
        print(json.dumps(error_output, indent=2), file=sys.stderr)
        sys.exit(1)

File: scripts/scaffolds/mlr/inference.js
/**
 * Multiple Linear Regression - Local Inference Script
 * Performs prediction using trained MLR model parameters.
 * Used for client-side visualization in the browser.
 * Input: parameters object, input data array
 * Output: Predictions array
 */

/**
 * Make predictions using the trained MLR model.
 * @param {Array<Object>} data - Array of data points with input features
 * @param {Object} parameters - Trained model parameters
 * @param {Array<number>} parameters.coefficients - Regression coefficients
 * @param {number} parameters.intercept - Regression intercept
 * @param {Array<string>} parameters.input_fields - Names of input fields
 * @param {string} parameters.output_field - Name of output field (default: 'y')
 * @returns {Array<Object>} Array of predictions with original data and predicted values
 */
function predict(data, parameters) {
  const { coefficients, intercept, input_fields, output_field = 'y' } = parameters;
  
  // Validate inputs
  if (!Array.isArray(coefficients) || typeof intercept !== 'number') {
    throw new Error('Invalid parameters: coefficients must be an array and intercept a number');
  }
  
  if (!Array.isArray(input_fields) || input_fields.length === 0) {
    throw new Error('Invalid parameters: input_fields must be a non-empty array');
  }
  
  if (coefficients.length !== input_fields.length) {
    throw new Error(`Coefficient count (${coefficients.length}) must match input field count (${input_fields.length})`);
  }
  
  if (!Array.isArray(data)) {
    throw new Error('Data must be an array');
  }
  
  // Make predictions
  const predictions = data.map(row => {
    // Extract feature values
    const features = input_fields.map(field => {
      const value = parseFloat(row[field]);
      if (isNaN(value)) {
        throw new Error(`Invalid value for field '${field}': ${row[field]}`);
      }
      return value;
    });
    
    // Calculate prediction: y = b0 + b1*x1 + b2*x2 + ... + bn*xn
    let y_pred = intercept;
    for (let i = 0; i < features.length; i++) {
      y_pred += coefficients[i] * features[i];
    }
    
    return {
      ...row,
      [`${output_field}_pred`]: y_pred
    };
  });
  
  return predictions;
}

/**
 * Calculate feature importance (contribution to predictions)
 * @param {Object} parameters - Model parameters
 * @param {Array<Object>} data - Sample data for analysis
 * @returns {Array<Object>} Feature importance rankings
 */
function calculateFeatureImportance(parameters, data) {
  const { coefficients, input_fields } = parameters;
  
  // Calculate mean absolute values for each feature
  const featureMeans = input_fields.map((field, i) => {
    const values = data.map(row => Math.abs(parseFloat(row[field]) || 0));
    const mean = values.reduce((sum, v) => sum + v, 0) / values.length;
    return mean;
  });
  
  // Calculate importance as |coefficient| * mean(|feature_values|)
  const importance = input_fields.map((field, i) => ({
    feature: field,
    coefficient: coefficients[i],
    importance: Math.abs(coefficients[i]) * featureMeans[i],
    percentage: 0  // Will be calculated after sorting
  }));
  
  // Sort by importance
  importance.sort((a, b) => b.importance - a.importance);
  
  // Calculate percentage of total importance
  const totalImportance = importance.reduce((sum, item) => sum + item.importance, 0);
  importance.forEach(item => {
    item.percentage = totalImportance > 0 ? (item.importance / totalImportance) * 100 : 0;
  });
  
  return importance;
}

/**
 * Generate prediction summary statistics
 * @param {Array<Object>} predictions - Predictions with actual and predicted values
 * @param {string} outputField - Name of the output field
 * @returns {Object} Summary statistics
 */
function getPredictionSummary(predictions, outputField = 'y') {
  const predField = `${outputField}_pred`;
  
  const values = predictions.map(p => p[predField]);
  const mean = values.reduce((sum, v) => sum + v, 0) / values.length;
  const min = Math.min(...values);
  const max = Math.max(...values);
  const variance = values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / values.length;
  const stdDev = Math.sqrt(variance);
  
  return {
    count: predictions.length,
    mean,
    min,
    max,
    stdDev,
    range: max - min
  };
}

// Export for use in browser (ES modules) or Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { 
    predict, 
    calculateFeatureImportance,
    getPredictionSummary 
  };
}

// Also support direct browser usage
if (typeof window !== 'undefined') {
  window.MLR_Inference = { 
    predict, 
    calculateFeatureImportance,
    getPredictionSummary 
  };
}
